File: BST.java
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);

		int qtd_bases, h_max, qtd_inicial;

		qtd_bases = in.nextInt();
		h_max = in.nextInt();
		qtd_inicial = in.nextInt();

		BinaryTree bases[] = new BinaryTree[qtd_bases];
		for(int i = 0; i < qtd_bases; i++) bases[i] = new BinaryTree();

		String comando;
		int casta, id, base;

		boolean stop = false;
		
		while(!stop) {
			comando = in.next();
			
			if(!comando.equals("END")) {
				casta = in.nextInt();
				id = in.nextInt();
				base = in.nextInt();

				Node temp = bases[base].root;

				if (comando.equals("INF") || comando.equals("ADM")) {

					int contador = 0;
					bases[base].insert(casta, id);

					if(bases[base].getDepthTree(temp) > h_max) {  //Checa se passou o limite de altura

						while(contador < qtd_bases) {
							bases[(base+contador)%qtd_bases].remove(temp, casta);
							contador++;

							temp = bases[(base+contador)%qtd_bases].root;

							bases[(base+contador)%qtd_bases].insert(casta, id);

							if(bases[(base+contador)%qtd_bases].getDepthTree(temp) <= h_max) break;						
						}
					}
					if(comando.equals("INF")) {
						if(contador>qtd_bases) System.out.println("-1");
						else System.out.println((base+contador)%qtd_bases);
					}
				} else if (comando.equals("EXT")) {
					bases[base].EXT(temp, casta, id);

				} else {
					System.out.println(bases[base].SCH(temp, casta, id));
				}	
			}
			else stop = true;
		}
	}
}		

class Postos {
	private int id;
	private Postos next;

	public Postos() {
		this.id = -1;
		this.next = null;
	}

	public void inserir(int id) {
		if (this.id == -1) {
			this.id = id;
			this.next = new Postos();
		} else {
			this.next.inserir(id);
		}
	}

	public boolean remover(int id) {
		if (this.id != -1) {
			if (this.id == id) {
				this.id = -2;
				return true;
			} else {
				return this.next.remover(id);
			}
		} else {
			return false;
		}
	}

	public boolean search(int id) {
		if (this.id != -1) {
			if (this.id == id) {
				return true;
			} else {
				return this.next.search(id);
			}
		} else {
			return false;
		}
	}

	public boolean isempty() { // checa se a remocao extinguiu o posto
		if (this.id != -1) {
			if (this.id == -2) {
				return this.next.isempty();
			} else {
				return false;
			}
		} else {
			return true;
		}
	}
}

class Node {
	int casta;
	Postos postos;
	Node left, right;

	public Node(int casta) {
		this.casta = casta;
		this.postos = new Postos();
		this.left = this.right = null;
	}
}

class BinaryTree {
	Node root;

	void insert(int casta, int id) { 
		root = insertRec(root, casta, id); 
	}
	public Node insertRec(Node root, int casta, int id) {
		if (root == null) { 
			root = new Node(casta);
			root.postos.inserir(id);
			return root; 
		} 	 
		if (casta < root.casta) 
			root.left = insertRec(root.left, casta, id); 
		else if (casta > root.casta) 
			root.right = insertRec(root.right, casta, id);
		else			//Achou a casta
			root.postos.inserir(id);
		return root; 
	}

	public void EXT(Node root, int casta, int id) {
		if (root != null) {
			if (root.casta == casta) { 		// Achou a casta de insercao
				root.postos.remover(id);
				if(root.postos.isempty()) {
					this.root = remove(this.root, casta);
				}
			} else if (casta > root.casta)
				EXT(root.right, casta, id);

			else
				EXT(root.left, casta, id);
		}
	}

	public int minCasta(Node root) {
		if (root.left == null)
			return root.casta;
		else
			return minCasta(root.left);
	}

	public Postos minPosto(Node root) {
		if (root.left == null)
			return root.postos;
		else
			return minPosto(root.left);
	}

	public Node remove(Node root, int casta) 
	{ 
		if (root == null)  return root; 
		if (casta < root.casta) 
			root.left = remove(root.left, casta); 
		else if (casta > root.casta) 
			root.right = remove(root.right, casta); 
		else
		{ 
			if (root.left == null) 
				return root.right; 
			else if (root.right == null) 
				return root.left; 
			else
				root.casta = minCasta(root.right);
			root.postos = minPosto(root.right);
			root.right = remove(root.right, root.casta); 
		} 
		return root; 
	}

	public int getDepthNode(Node root, int casta, int level) {
		if (root == null)
			return 0;

		if (root.casta == casta)
			return level;

		int downlevel = getDepthNode(root.left, casta, level + 1);
		if (downlevel != 0)
			return downlevel;

		else downlevel = getDepthNode(root.right, casta, level + 1);
		return downlevel;
	}

	public int SCH(Node root, int casta, int id) {
		if (root != null) {
			if (root.casta == casta) { 		// Achou a casta

				if (root.postos.search(id)) 
					return getDepthNode(this.root, casta, 0);
				else
					return -1;

			} else if (casta > root.casta)
				return SCH(root.right, casta, id);
			else
				return SCH(root.left, casta, id);
		} else
			return -1;
	}

	public int getDepthTree(Node root) {
		if (root == null)
			return 0;
		else {
			int lDepth = getDepthTree(root.left);
			int rDepth = getDepthTree(root.right);

			if (lDepth > rDepth)
				return (lDepth + 1);
			else
				return (rDepth + 1);
		}
	}
}

File: BST.cpp
#include <bits/stdc++.h>
using namespace std;

struct NodeList {
    int valor;
    NodeList *next = NULL;
};

struct Posto{
    NodeList* head = NULL;

    void insert(int id){
        NodeList *ptr = new NodeList();
        ptr->valor = id;
        ptr->next = head;
        head = ptr;
    }
    bool search(int id){
        NodeList *current = head;
        while(current != NULL){
            if(current->valor == id) return true;
            else current = current->next;
        }
        return false;
    }
    void remove(int v){
        if(search(v) == true){
            NodeList* temp = head;
            NodeList* prev = NULL;

            if (temp != NULL && temp->valor == v)
            {
                head = temp->next;
                return;
            }
            while (temp != NULL && temp->valor != v)
            {
                prev = temp;
                temp = temp->next;
            }
            prev->next = temp->next;
        }
    }
    bool isEmpty(){
        if(head == NULL) return true;
        else return false;
    }
};

struct NodeTree{
    int casta;
    Posto posto;
    NodeTree* left;
    NodeTree* right;
};

struct Tree{
    NodeTree* root = NULL;

    NodeTree* newNode(int casta, int id){
        NodeTree *temp = new NodeTree();
        temp->casta = casta;
        temp->left = temp->right = NULL;
        temp->posto.insert(id);
        return temp;
    };
    NodeTree* insertNode(NodeTree* root, int casta, int id){
        if(root == NULL) return newNode(casta, id);
        else if(casta < root->casta) root->left = insertNode(root->left, casta, id);
        else root->right = insertNode(root->right, casta, id);

        return root;
    };
    NodeTree* search(NodeTree* root, int casta)
    {
        if (root == NULL || root->casta == casta)
            return root;

        if (root->casta < casta)
            return search(root->right, casta);

        return search(root->left, casta);
    }
    int getHeight(NodeTree *node, int data, int level)
    {
        if (node == NULL)
            return 0;

        if (node->casta == data)
            return level;

        int downlevel = getHeight(node -> left, data, level + 1);
        if (downlevel != 0)
            return downlevel;

        downlevel = getHeight(node->right, data, level + 1);
        return downlevel;
    }

    int maxDepth(NodeTree* node)
    {
        if (node == NULL)
            return 0;
        else
        {
            int lDepth = maxDepth(node->left);
            int rDepth = maxDepth(node->right);

            if (lDepth > rDepth)
                return(lDepth + 1);
            else return(rDepth + 1);
        }
    }

    NodeTree * minValueNode(NodeTree* node)
    {
        NodeTree* current = node;
        while (current && current->left != NULL)
            current = current->left;

        return current;
    }

    NodeTree* deleteNode(NodeTree* root, int key)
    {
        if (root == NULL) return root;

        if (key < root->casta)
            root->left = deleteNode(root->left, key);

        else if (key > root->casta)
            root->right = deleteNode(root->right, key);

        else
        {
            if (root->left == NULL)
            {
                NodeTree *temp = root->right;
                free(root);
                return temp;
            }
            else if (root->right == NULL)
            {
                NodeTree *temp = root->left;
                free(root);
                return temp;
            }

            NodeTree* temp = minValueNode(root->right);

            root->casta = temp->casta;
            root->posto = temp->posto;

            root->right = deleteNode(root->right, temp->casta);
        }
        return root;
    }
};

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, h, f;
    cin >> n >> h >> f;

    Tree* bases[n];
    for(int i = 0 ; i < n; i++) bases[i] = new Tree();

    string comando;
    cin >> comando;

    while(comando != "END"){

        int c, id, b;
        cin >> c >> id >> b;

        if(comando == "ADM" || comando == "INF"){
            NodeTree* temp = bases[b]->root;
            NodeTree* aux = bases[b]->search(temp, c);

            int contador = 0;
            if(aux == NULL){
                bases[b]->root = bases[b]->insertNode(temp, c, id);

                if(bases[b]->maxDepth(temp) > h) {  //Checa se passou o limite de altura
                    while(contador < n) {
                        bases[(b+contador)%n]->deleteNode(temp, c);
                        contador++;

                        temp = bases[(b+contador)%n]->root;

                        aux = bases[b]->search(temp, c);
                        if(aux == NULL) {
                            temp = bases[(b+contador)%n]->insertNode(temp, c, id);
                            if(bases[(b+contador)%n]->maxDepth(temp) <= h)
                                break;
                        } else{
                            aux->posto.insert(id);
                            break;
                        }
                    }
                }
            }
            else
                aux->posto.insert(id);

            if(comando == "INF") {
                if(contador > n) cout << -1 << endl;
                else cout << (b+contador)%n << endl;
            }
        }
        else if(comando == "SCH" || comando == "EXT"){
            NodeTree* temp = bases[b]->search(bases[b]->root, c);

            if(temp != NULL){
                if(temp->posto.search(id)){
                    if(comando == "SCH")
                        cout << bases[b]->getHeight(bases[b]->root, c, 0) << endl;
                    else {
                        temp->posto.remove(id);
                        if(temp->posto.isEmpty())
                            bases[b]->deleteNode(bases[b]->root, c);
                    }
                }
                else if (comando == "SCH")
                    cout << -1 << endl;
            }
            else if(comando == "SCH")
                cout << -1 << endl;
        }
        cin >> comando;
    }
}
